src/Umbraco.Web/umbraco.presentation/content.cs

```diff
@@ -489,119 +488,71 @@ private XmlDocument LoadContentFromDatabase()
+                        var doctype = xmlDoc.CreateDocumentType("root", null, null, 
+                            ApplicationContext.Current.Services.ContentTypeService.GetContentTypesDtd());
+                        xmlDoc.AppendChild(doctype);
+                        var parent = xmlDoc.CreateElement("root");
+                        var pIdAtt = xmlDoc.CreateAttribute("id");
+                        pIdAtt.Value = "-1";
+                        parent.Attributes.Append(pIdAtt);
+                        xmlDoc.AppendChild(parent);
 
                         // Esben Carlsen: At some point we really need to put all data access into to a tier of its own.
                         // CLN - added checks that document xml is for a document that is actually published.
-                        string sql =
-                            @"select umbracoNode.id, umbracoNode.parentId, umbracoNode.sortOrder, cmsContentXml.xml from umbracoNode 
+                        const string sql = @"select umbracoNode.id, umbracoNode.parentID, umbracoNode.sortOrder, cmsContentXml.xml, umbracoNode.level from umbracoNode
 inner join cmsContentXml on cmsContentXml.nodeId = umbracoNode.id and umbracoNode.nodeObjectType = @type
 where umbracoNode.id in (select cmsDocument.nodeId from cmsDocument where cmsDocument.published = 1)
-order by umbracoNode.level, umbracoNode.sortOrder";
-
-
+order by umbracoNode.level, umbracoNode.parentID, umbracoNode.sortOrder";
 
-                        using (
-                            IRecordsReader dr = SqlHelper.ExecuteReader(sql,
-                                                                        SqlHelper.CreateParameter("@type",
-                                                                                                  new Guid(
-                                                                                                      Constants.ObjectTypes.Document)))
-                            )
+                        XmlElement last = null;
+                        var db = ApplicationContext.Current.DatabaseContext.Database;
+                        //NOTE: Query creates a reader - does not load all into memory
+                        foreach (var row in db.Query<dynamic>(sql, new { type  = new Guid(Constants.ObjectTypes.Document)}))
                         {
-                            while (dr.Read())
+                            string parentId = ((int)row.parentID).ToInvariantString();
+                            string xml = row.xml;
+
+                            //if the parentid is changing
+                            if (last != null && last.GetAttribute("parentID") != parentId)
                             {
-                                int currentId = dr.GetInt("id");
-                                int parentId = dr.GetInt("parentId");
-                                string xml = dr.GetString("xml");
-
-                                // fix sortOrder - see notes in UpdateSortOrder

```

Comment:
Are you sure we can get rid of that fixing? It was for cases where the xml in database would have an invalid sort order, because in some cases (when reordering) the xml is not updated.